clang/lib/AST/DeclarationName.cpp

```diff
@@ -301,12 +302,21 @@ DeclarationNameTable::getCXXDeductionGuideName(TemplateDecl *Template) {
   return DeclarationName(Name);
 }
 
+namespace clang {
+  void *getUniqueOpaquePtr(QualType Ty) {
+    if (auto Record = Ty->getAsCXXRecordDecl()) {
+      return Record;
+    }
+    return Ty.getAsOpaquePtr();
```

Comment:
I would try doing something like `Ty.dump();`, especially if you have a small unit test, just to see what's getting through.